src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,332 @@
+            PutDepositFor(engine, to, deposit);
+        }
+
+        /// <summary>
+        /// Lock asset until the specified height is unlocked.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="addr">Account</param>
+        /// <param name="till">specified height</param>
+        /// <returns>Whether deposit lock height was successfully updated.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
+            if (till < Ledger.CurrentIndex(engine.SnapshotCache) + 2) return false; // deposit must be valid at least until the next block after persisting block.
+            Deposit deposit = GetDepositFor(engine.SnapshotCache, addr);
+            if (deposit is null || till < deposit.Till) return false;
+            deposit.Till = till;
+
+            PutDepositFor(engine, addr, deposit);
+            return true;
+        }
+
+        /// <summary>
+        /// ExpirationOf returns deposit lock height for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit lock height of the specified address.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public uint ExpirationOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Till;
+        }
+
+        /// <summary>
+        /// BalanceOf returns deposited GAS amount for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit balance of the specified account.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public BigInteger BalanceOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Amount;
+        }
+
+        /// <summary>
+        /// Withdraw sends all deposited GAS for "from" address to "to" address. If "to"
+        /// address is not specified, then "from" will be used as a sender.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">From Account</param>
+        /// <param name="to">To Account</param>
+        /// <returns>Whether withdrawal was successfull.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private async ContractTask<bool> Withdraw(ApplicationEngine engine, UInt160 from, UInt160 to)
```

Comment:
Never return false?

Reply:
Originally, it returned `false` in several cases:
https://github.com/nspcc-dev/neo-go/blob/553db45fef26e692507dce813a6b1fdfd12fec77/pkg/core/native/notary.go#L296
https://github.com/nspcc-dev/neo-go/blob/553db45fef26e692507dce813a6b1fdfd12fec77/pkg/core/native/notary.go#L304
https://github.com/nspcc-dev/neo-go/blob/553db45fef26e692507dce813a6b1fdfd12fec77/pkg/core/native/notary.go#L308

But then changes was requested and we converted these cases to exceptions in C# implementation (Go implementation will follow if we merge this PR). So I think we firstly need to agree on the final desired behaviour. What is a preferred behaviour for you? And also for @roman-khimov?

Reply:
It was made to follow the behavior of `transfer()` which returns `false` routinely in similar error cases. It's a bit more flexible and I'd suggest doing that instead of the current code. Also notice that changing API is not always possible at this stage for compatibility reasons. Although in this particular case of `withdraw()` it probably can be done because I doubt we've ever called this method.

Reply:
Fixed.